master/internal/api_token.go

```diff
@@ -0,0 +1,215 @@
+// access token for the given user.
+func (a *apiServer) PostAccessToken(
+	ctx context.Context, req *apiv1.PostAccessTokenRequest,
+) (*apiv1.PostAccessTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	targetFullUser, err := getFullModelUser(ctx, model.UserID(req.UserId))
+	if err != nil {
+		return nil, err
+	}
+	targetUser := targetFullUser.ToUser()
+	if err = token.AuthZProvider.Get().CanCreateAccessToken(ctx, *curUser, targetUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenExpiration := token.DefaultTokenLifespan
+	if req.Lifespan != "" {
+		d, err := time.ParseDuration(req.Lifespan)
+		if err != nil || d < 0 {
+			return nil, status.Error(codes.InvalidArgument,
+				"Lifespan must be a Go-formatted duration string with a positive value")
+		}
+		tokenExpiration = d
+	}
+
+	token, tokenID, err := token.CreateAccessToken(
+		ctx, targetFullUser.ID, token.WithTokenExpiry(&tokenExpiration), token.WithTokenDescription(req.Description))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.PostAccessTokenResponse{Token: token, TokenId: int32(tokenID)}, nil
+}
+
+type accessTokenFilter struct {
+	OnlyActive bool            `json:"only_active"`
+	Username   string          `json:"username"`
+	TokenIDs   []model.TokenID `json:"token_ids"`
+}
+
+// GetAccessTokens returns all access token info.
+func (a *apiServer) GetAccessTokens(
+	ctx context.Context, req *apiv1.GetAccessTokensRequest,
+) (*apiv1.GetAccessTokensResponse, error) {
+	sortColMap := map[apiv1.GetAccessTokensRequest_SortBy]string{
+		apiv1.GetAccessTokensRequest_SORT_BY_UNSPECIFIED: "id",
+		apiv1.GetAccessTokensRequest_SORT_BY_USER_ID:     "user_id",
+		apiv1.GetAccessTokensRequest_SORT_BY_EXPIRY:      "expiry",
+		apiv1.GetAccessTokensRequest_SORT_BY_CREATED_AT:  "created_at",
+		apiv1.GetAccessTokensRequest_SORT_BY_TOKEN_TYPE:  "token_type",
+		apiv1.GetAccessTokensRequest_SORT_BY_REVOKED:     "revoked",
+		apiv1.GetAccessTokensRequest_SORT_BY_DESCRIPTION: "description",
+	}
+	orderByMap := map[apiv1.OrderBy]string{
+		apiv1.OrderBy_ORDER_BY_UNSPECIFIED: "ASC",
+		apiv1.OrderBy_ORDER_BY_ASC:         "ASC",
+		apiv1.OrderBy_ORDER_BY_DESC:        "DESC",
+	}
+
+	userSessions := []model.UserSession{}
+
+	query := db.Bun().NewSelect().Model(&userSessions).
+		ModelTableExpr("user_sessions as us").
+		Column("us.id").
+		Column("us.user_id").
+		Column("us.expiry").
+		Column("us.created_at").
+		Column("us.token_type").
+		Column("us.revoked").
+		Column("us.description")
+
+	var userIDForGivenUsername model.UserID
+	if req.Filter != "" {
+		var atf accessTokenFilter
+		err := json.Unmarshal([]byte(req.Filter), &atf)
+		if err != nil {
+			return nil, err
+		}
+
+		if atf.Username != "" {
+			nameFilterExpr := "%" + atf.Username + "%"
+			err := db.Bun().NewSelect().
+				Table("users").
+				Column("id").
+				Where("username ILIKE ?", nameFilterExpr).
```

Comment:
hold up. what's going on here? am i missing something, why are you doing a fuzzy/pattern search? 

would this not match and return all user token that contain the string passed in? if so, huge security problem.

Reply:
I agree we can check directly, and if thats the better way, we can update this. However, even if this returns multiple userIds, we will be checking the permissions of the requestor before we query the table for token info.
So I do not think this will create the security problem, but it makes sense to check equality instead of `ILIKE` pattern match.

Reply:
ah, ok, i thought we were doing the security check at the top of this function, since usually it's the very first thing we do in API methods. so not a security problem, but this is definitely a bug.

1. i do think we should do the perms/auth check at the top of this function. like check that the request's user has access to the functionality they are requesting. and if they don't, just exit immediately. it prevents someone from adding additional functionality before the perms check that may leak unintended info (like in error logs or something).

2. i can't think of a reason why we'd want to do pattern matching at all here, but maybe i'm missing something. have you tested this against lots of users? i.e. request username = "a" in a db with many users. seems like you'd get an error if you searched for any username that was a substring of another username, because you're scanning it into a singular `model.UserId`. also, if user "a" didn't exist but "abc" existed, the query would return it.

Reply:
1. permission check is done in line 158 - 161 before we actually run the scan of the Bun query. We will not be able to bring in the top of the API functionality as we need to check the permissions of current user and target user (getting from the filter part of the query)
2. I will be removing the pattern matching, instead keep complete matching of the usernames

Reply:
makes sense yeah, i think i'm proposing something like: 
- query for requested user ID from request username
- do perms check on current user & requested user ID
- continue with rest of token fetch logic